predicators/envs/narrow_passage.py

```diff
@@ -0,0 +1,490 @@
+                                 np.array([target_x, target_y]))
+        return result
+
+    def _MoveToTarget_terminal(self, state: State, memory: Dict,
+                               objects: Sequence[Object],
+                               params: Array) -> bool:
+        del memory, params  # unused
+        return self._TouchedGoal_holds(state, objects)
+
+    @staticmethod
+    def _MoveAndOpenDoor_policy(state: State, memory: Dict,
+                                objects: Sequence[Object],
+                                params: Array) -> Action:
+        del state, objects, params  # unused
+        assert memory["action_plan"], "Motion plan did not reach its goal"
+        return memory["action_plan"].pop(0)
+
+    def _MoveAndOpenDoor_initiable(self, state: State, memory: Dict,
+                                   objects: Sequence[Object],
+                                   params: Array) -> bool:
+        robot, door = objects
+        # If door is already open, just return True because the action is done
+        if self._DoorIsOpen_holds(state, [door]):
+            # set action plan to nothing
+            memory["action_plan"] = [
+                Action(np.array([0.0, 0.0, 0.0], dtype=np.float32))
+            ]
+            return True
+        if self._robot_near_door(state):
+            # if robot is already within range of the door, just open the door
+            memory["action_plan"] = [
+                Action(np.array([0.0, 0.0, 1.0], dtype=np.float32))
+            ]
+            return True
+        # Select target point slightly above door
+        door_center_x = state.get(door, "x")
+        door_target_y = (
+            self.y_ub - self.y_lb
+        ) / 2 + self.y_lb + self.door_sensor_radius - self.robot_radius
+        result = self._run_birrt(state, memory, params, robot,
+                                 np.array([door_center_x, door_target_y]))
+        if result is False:
+            # Failed to find motion plan, so option is not initiable
+            return False
+        # Append open door action to memory action plan
+        memory["action_plan"].append(
+            Action(np.array([0.0, 0.0, 1.0], dtype=np.float32)))
+        return True
+
+    def _MoveAndOpenDoor_terminal(self, state: State, memory: Dict,
+                                  objects: Sequence[Object],
+                                  params: Array) -> bool:
+        del memory, params  # unused
+        return self._DoorIsOpen_holds(state, objects[1:])
+
+    def _run_birrt(self, state: State, memory: Dict, params: Array,
+                   robot: Object, target_position: Array) -> bool:
+        """Runs BiRRT to motion plan from start to target positions, and store
+        the position and action plans in memory if successful.
+
+        Returns true if successful, else false
+        """
+        # Make a BiRRT object and return it for use
+        # The seed is determined by the parameter passed into the option.
```

Comment:
add a note that this is a hack to prevent bilevel planning from giving up after one attempt, which it would do if the option had an empty params space